Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -687,6 +687,82 @@ public struct NonBlockingFileIO: Sendable {
         }
     }
 
+    private func createDirectory0(_ path: String, mode: NIOPOSIXFileMode) throws {
+        let pathView = path.utf8
+        if pathView.isEmpty {
+            return
+        }
+
+        // Fail fast if not a directory or file exists
+        do {
+            var s = stat()
+            try Posix.stat(pathname: path, outStat: &s)
+            if (S_IFMT & s.st_mode) == S_IFDIR {
+                return
+            }
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch let error as IOError where error.errnoCode == ENOENT {
+            // if directory does not exist we can proceed with creating it
+        }
+
+        // Slow path, check that all intermediate directories exist recursively
+
+        // Trim any trailing path separators
+        var index = pathView.index(before: pathView.endIndex)
+        let pathSeparator = 47 // 47 is `/`
+        while index != pathView.startIndex && pathView[index] == pathSeparator {
+            index = pathView.index(before: index)
+        }
+
+        // Find first non-trailing path separator if it exists
+        while index != pathView.startIndex && pathView[index] != pathSeparator {
+            index = pathView.index(before: index)
+        }
+
+        // If non-trailing path separator is found, create parent directory
+        if index > pathView.startIndex {
+            try self.createDirectory0(String(decoding: pathView.prefix(upTo: index), as: UTF8.self), mode: mode)
```

Comment:
Nit: I _think_ you can create a `Substring` from a `String.UTF8View.Subsequence`, and then create a `String`, so you should be able to do:

```
try self.createDirectory0(String(Substring(pathView.prefix(upTo: index))), mode: mode)
```

Reply:
ah, fixed